### PR TITLE
Security

### PR DESCRIPTION
**Pull request by pwnfooj716**
Check over my code and possibly test it out also if you want. To prevent brute-force attacks, I keep track of failed attempts and lock the account for an exponentially increasing amount of time. The time will become noticeable at about 10 failed attempts which will lock the account for approx 10 seconds. At that point, you should see an error message on the login page explaining the account locking.

### CODE REVIEW

File: routes/login.js [hunk 1]
```diff
@@ -9,11 +10,41 @@ router.get("/", (req, res) => {
...
+    passport.authenticate('local', (err, user, info) => {
+        if (err) {
+            return next(err);
+        }
+        return db.accountLocked(req.body.username).then((lockData) => {
+            if (lockData.locked) {
+                let date = new Date(lockData.lockAccountUntil);
+                let data = {
+                    hasErrors: true,
+                    errors: [`${lockData.failedLoginAttempts} failed login attempts. Your account has been locked until ${date}`]
```
Comment (avpchamp) [inline routes/login.js:25]: does this error display?

File: data/index.js [hunk 2]
```diff
@@ -186,6 +186,67 @@ function getAssignment(assignmentId) {
...
+			});
+		});
+	},
+	// User: Passport
+	accountLocked(username) {
+		let lockData = {};
+		return students().then((collection) => {
+			return collection.findOne({username: username}).then((student) => {
+				if (student) {
+					lockData.failedLoginAttempts = student.failedLoginAttempts,
```
Comment (avpchamp) [inline data/index.js:216]: does this still work if student.failedLoginAttempts is undefined

Comment (avpchamp) [reply data/index.js:216]: same with others below

Comment (avpchamp) [reply data/index.js:216]: oh wait nvm you add field when creating student

File: routes/login.js [hunk 3]
```diff
@@ -9,11 +10,41 @@ router.get("/", (req, res) => {
...
+                };
+                return res.render('login/login', data);
+            }
+            if (!user) {
+                // Login failed
+                return db.failedLoginAttempt(req.body.username, lockData.failedLoginAttempts).then(() => {
+                    return res.redirect('/login');
```
Comment (avpchamp) [inline routes/login.js:32]: should we send the user an error message that there login failed

File: routes/login.js (hunk 1 above)
Comment (pwnfooj716) [reply routes/login.js:25]: Yes. You need to fail the password about 10 times in a row in order for the time delay to be big enough for your attempt to actually be denied. At this point, you should start seeing the message.

File: routes/login.js (hunk 3 above)
Comment (pwnfooj716) [reply routes/login.js:32]: I send the error iff their actual attempt was blocked. You will not see the error message even if you fail the attempt, but enough time has passed for it not to be denied. I tried showing the message every time the user failed the login but it looked a bit obnoxious.

**Review by avpchamp — APPROVED** [review]

**Pull request merged**